xls/contrib/integrator/integration_algorithms/basic_integration_algorithm.h

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// A abstract class for an algorithm to integrate multiple functions.
+
+#ifndef XLS_INTEGRATOR_BASIC_INTEGRATION_ALGORITHM_
+#define XLS_INTEGRATOR_BASIC_INTEGRATION_ALGORITHM_
+
+#include "xls/contrib/integrator/integration_algorithms/integration_algorithm.h"
+
+namespace xls {
+
+// A naive merging algorithm.  A node is eligible to be added to the
+// integration function when all of its operands have already been added.
+// At each step, adds the eligible node for which the cost of adding it to
+// the function (either by inserting or merging with any integeration function
+// node) is the lowest.
+class BasicIntegrationAlgorithm
+    : public IntegrationAlgorithm<BasicIntegrationAlgorithm> {
+ public:
+  BasicIntegrationAlgorithm(const BasicIntegrationAlgorithm& other) = delete;
+  void operator=(const BasicIntegrationAlgorithm& other) = delete;
+
+ private:
+  // IntegrationAlgorithm::IntegrateFunctions needs to be able to call derived
+  // class constructor.
+  friend class IntegrationAlgorithm;
+
+  BasicIntegrationAlgorithm(
+      Package* package, absl::Span<const Function* const> source_functions,
+      const IntegrationOptions& options = IntegrationOptions())
+      : IntegrationAlgorithm(package, source_functions, options) {}
+
+  // Represents a possible modification to the integration function.
+  struct BasicIntegrationMove : IntegrationMove {
+    std::list<Node*>::iterator node_itr;
+  };
+
+  // Make a BasicIntegrationMove for an insert.
+  inline BasicIntegrationMove MakeInsertMove(
+      std::list<Node*>::iterator node_itr, int64 cost) {
+    return BasicIntegrationMove{{.node = *node_itr,
+                                 .move_type = IntegrationMoveType::kInsert,
+                                 .cost = cost},
+                                .node_itr = node_itr};
+  }
+
+  // Make a BasicIntegrationMove for a merge.
+  inline BasicIntegrationMove MakeMergeMove(std::list<Node*>::iterator node_itr,
+                                            Node* merge_node, int64 cost) {
+    return BasicIntegrationMove{{.node = *node_itr,
+                                 .move_type = IntegrationMoveType::kMerge,
+                                 .merge_node = merge_node,
+                                 .cost = cost},
+                                .node_itr = node_itr};
```

Comment:
Designated initializers are a C++20 extension, suggest not using for greater compatibility.

If used, then code should not mix designated and non-designated, nor nest designated initializers.

ex.
```
    return BasicIntegrationMove{{.node = *node_itr,
                                 .move_type = IntegrationMoveType::kInsert,
                                 .cost = cost},
                                 node_itr};
```

Reply:
Ack.

"Designated initializers are a C++20 extension, suggest not using for greater compatibility."
Designated initializers are already in use in multple parts of XLS

"If used, then code should not mix designated and non-designated, nor nest designated initializers."
I'm a little unclear on what specifically in the initializations this is refering to - I'm assuming this inner bracketed component?
`{.node = *node_itr,
  .move_type = IntegrationMoveType::kInsert,
  .cost = cost}`

Maybe the issue is that this looks like its own designated intializer, with the resulting struct being passed into the outer designated initializer without a designation  / label?
The designated initialzers here are creating BasicIntegrationMove structs.  This struct inherets from the IntegrationMove struct which has fields node, move_type, merge_node, and cost.  So here, we are actually assigning the node, move_type, merge_node, and cost fields of the BasicIntegrationMove struct with designations.  The extra brackets are necessary because these fields are inhereted.

Reply:
Issue is the outer bracketed component.  The first is using a non-designated initialization, while the second is using the designated initialization of .node_itr.  Suggestion as commented above -- remove the designated initialization of the outer bracket.

Reply:
Done.